Arriba/Arriba.Web/parts/Search.jsx

```diff
@@ -53,60 +49,75 @@ export default React.createClass({
             }).cleaned);
         }
 
-        return Object.assign(this.getEmptyState(), {
-            allBasics: [],
+        this.state = {
             tables: [],
             page: 0,
-            hasMoreData: false,
             query: this.props.params.q || "",
             currentTable: table,
             currentTableSettings: {}, // {} denote no state, do not set to null.
             userSelectedTable: table,
             userSelectedId: this.props.params.open
-        });
-    },
-    componentDidMount: function () {
-        window.addEventListener("beforeunload", this); // For Mru
-        this.mru = new Mru();
+        };
+
+        this.events = {
+            "beforeunload": e => this.mru.push(),
+            "storage": e => { if (e.key.startsWith("table-")) this.getTableSettings() },
+        };
+    }
+    componentDidMount() {
+        super.componentDidMount();
         this.componentDidUpdate({}, {});
-    },
-    componentDidUpdate: function(prevProps, prevState) {
+    }
+    componentDidUpdate(prevProps, prevState) {
         const diffProps = Object.diff(prevProps, this.props);
-        const diff = Object.diff(prevState, this.state);
+        const diffState = Object.diff(prevState, this.state);
+
+        if (diffProps.hasAny("allBasics") || diffState.hasAny("debouncedQuery")) {
+            this.getCounts();
+        }
+
+        if (diffState.hasAny("userSelectedTable", "counts")) {
+            const currentTable = this.state.userSelectedTable || this.state.counts && this.state.counts.resultsPerTable[0].tableName;
+            this.setState({ currentTable: currentTable });
+        }
+
+        // Do not wipe userSelectedId if currentTable is going from `undefined` to defined.
+        // Scnario: On page load with open=something and table inferred from query.
+        if (diffState.hasAny("currentTable") && prevState.currentTable) {
+            this.setState({ userSelectedId: undefined });
+        }
+
+        if (diffProps.hasAny("allBasics") || diffState.hasAny("currentTable")) {
+            this.getTableSettings();
+        }
 
-        if (diffProps.hasAny("allBasics")) {
-            this.getAllCounts();
+        if (diffState.hasAny("debouncedQuery", "currentTableSettings", "page")) {
+            this.getListings();
         }
 
         // Watching currentTable as sometimes inferred from the query.
         // Not watching for query changes (to match old behavior).
-        if (diffProps.hasAny("allBasics") || diff.hasAny("currentTable", "userSelectedId")) {
+        if (diffProps.hasAny("allBasics") || diffState.hasAny("currentTable", "userSelectedId")) {
             this.getDetails();
         }
 
-        if (diff.hasAny("query", "userSelectedTable", "userTableSettings", "userSelectedId", "currentTable")) {
-            this.setHistory();
+        if (diffState.hasAny("query", "userSelectedTable", "userTableSettings", "userSelectedId", "currentTable")) {

```

Comment:
These all look right except currentTable - we don't need to persist the table if the user hasn't explicitly selected one, right?

Reply:
If the user modifies the column list and sorting, then we have been persisting the current table to the Url even if the table was not explicitly selected.